File: src/CommandLine/ForEachVarHelpers.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

public static class ForEachVarHelpers
{
    public static List<Command> ExpandForEachVars(List<Command> commands)
    {
        return commands.SelectMany(command => command is ChatCommand chatCommand
            ? ExpandChatCommand(chatCommand)
            : new List<Command> { command }).ToList();
    }

    public static ForEachVariable ParseForeachVarOption(string[] args, out int skipCount)
    {
        skipCount = 0;

        var badLength = args.Length < 4;
        if (badLength) throw new CommandLineException($"Invalid foreach variable format. Expected: 'var name in value1 value2...'");

        var missingVarKeyword = args[0] != "var";
        if (missingVarKeyword) throw new CommandLineException($"Invalid foreach variable format. Expected: 'var name in value1 value2...'");

        var varName = args[1];

        var missingInKeyword = args[2] != "in";
        if (missingInKeyword) throw new CommandLineException($"Invalid foreach variable format. Expected: 'var name in value1 value2...'");

        skipCount += args.Length;
        var variableArgs = args.Skip(3).ToArray();

        var treatLinesAsValues = variableArgs.Length == 1 && variableArgs[0].Contains("\n");
        var values = treatLinesAsValues
            ? variableArgs[0].Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
            : variableArgs;

        int start = 0, end = 0;
        var treatAsIntRangeSeparatedByTwoDots = values.Length == 1 && IsIntRangeSeparatedByTwoDots(values[0], out start, out end);
        if (treatAsIntRangeSeparatedByTwoDots)
        {
            values = Enumerable.Range(start, end - start + 1).Select(i => i.ToString()).ToArray();
        }

        return new ForEachVariable(varName, values.ToList());
    }

    private static bool IsIntRangeSeparatedByTwoDots(string v, out int start, out int end)
    {
        start = end = 0;

        var parts = v.Split("..");
        if (parts.Length != 2) return false;

        start = int.Parse(parts[0]);
        end = int.Parse(parts[1]);
        if (start > end) return false;

        return true;
    }

    private static IEnumerable<Command> ExpandChatCommand(ChatCommand chatCommand)
    {
        var foreachVars = chatCommand.ForEachVariables;
        if (foreachVars.Count == 0)
        {
            return new List<ChatCommand> { chatCommand };
        }

        // Generate all combinations of variable values (Cartesian product)
        var combinations = GenerateValueCombinations(foreachVars);
        
        // Create a new command for each combination
        var expandedCommands = new List<ChatCommand>();
        foreach (var combination in combinations)
        {
            var clonedCommand = chatCommand.Clone();
            
            // Set variables for this combination
            for (int i = 0; i < foreachVars.Count; i++)
            {
                var varName = foreachVars[i].Name;
                var varValue = combination[i];
                clonedCommand.Variables[varName] = varValue;
                
                // Also set in ConfigStore so it's available for any templating
                ConfigStore.Instance.SetFromCommandLine($"Var.{varName}", varValue);
            }
            
            expandedCommands.Add(clonedCommand);
        }
        
        return expandedCommands;
    }

    private static List<List<string>> GenerateValueCombinations(List<ForEachVariable> foreachVars)
    {
        // Start with a single empty combination
        var combinations = new List<List<string>> { new List<string>() };
        
        // For each variable, extend all existing combinations with each value of the variable
        foreach (var foreachVar in foreachVars)
        {
            var newCombinations = new List<List<string>>();
            
            foreach (var existingCombination in combinations)
            {
                foreach (var value in foreachVar.Values)
                {
                    var newCombination = new List<string>(existingCombination);
                    newCombination.Add(value);
                    newCombinations.Add(newCombination);
                }
            }
            
            combinations = newCombinations;
        }
        
        return combinations;
    }
}

File: src/commonjs/src/CommandLine/ForEachVarHelpers.ts
import { Command } from './Command';
import { CommandWithVariables } from './CommandWithVariables';
import { ForEachVariable } from './ForEachVariable';
import { CommandLineException } from './CommandLineException';

export class ForEachVarHelpers {
    static expandForEachVars(commands: Command[]): Command[] {
        return commands.flatMap(command => 
            command instanceof CommandWithVariables
                ? this.expandCommandWithVars(command)
                : [command]
        );
    }

    static parseForeachVarOption(args: string[]): { variable: ForEachVariable; skipCount: number } {
        let skipCount = 0;

        if (args.length < 4) {
            throw new CommandLineException("Invalid foreach variable format. Expected: 'var name in value1 value2...'");
        }

        if (args[0] !== "var") {
            throw new CommandLineException("Invalid foreach variable format. Expected: 'var name in value1 value2...'");
        }

        const varName = args[1];

        if (args[2] !== "in") {
            throw new CommandLineException("Invalid foreach variable format. Expected: 'var name in value1 value2...'");
        }

        skipCount += args.length;
        const variableArgs = args.slice(3);

        const treatLinesAsValues = variableArgs.length === 1 && variableArgs[0].includes('\n');
        let values: string[];
        
        if (treatLinesAsValues) {
            values = variableArgs[0].split(/[\r\n]/).filter(line => line.trim() !== '');
        } else {
            values = variableArgs;
        }

        // Check for integer range (e.g., "1..10")
        if (values.length === 1) {
            const rangeMatch = this.parseIntRange(values[0]);
            if (rangeMatch) {
                const { start, end } = rangeMatch;
                values = Array.from({ length: end - start + 1 }, (_, i) => (start + i).toString());
            }
        }

        return {
            variable: new ForEachVariable(varName, values),
            skipCount
        };
    }

    private static parseIntRange(value: string): { start: number; end: number } | null {
        const parts = value.split('..');
        if (parts.length !== 2) return null;

        const start = parseInt(parts[0], 10);
        const end = parseInt(parts[1], 10);
        
        if (isNaN(start) || isNaN(end) || start > end) {
            return null;
        }

        return { start, end };
    }

    private static expandCommandWithVars(commandWithVars: CommandWithVariables): CommandWithVariables[] {
        const foreachVars = commandWithVars.forEachVariables;
        if (foreachVars.length === 0) {
            return [commandWithVars];
        }

        // Generate all combinations of variable values (Cartesian product)
        const combinations = this.generateValueCombinations(foreachVars);
        
        // Create a new command for each combination
        const expandedCommands: CommandWithVariables[] = [];
        
        for (const combination of combinations) {
            const clonedCommand = commandWithVars.clone();
            
            // Set variables for this combination
            for (let i = 0; i < foreachVars.length; i++) {
                const varName = foreachVars[i].name;
                const varValue = combination[i];
                clonedCommand.variables.set(varName, varValue);
            }
            
            expandedCommands.push(clonedCommand);
        }
        
        return expandedCommands;
    }

    private static generateValueCombinations(foreachVars: ForEachVariable[]): string[][] {
        // Start with a single empty combination
        let combinations: string[][] = [[]];
        
        // For each variable, extend all existing combinations with each value of the variable
        for (const foreachVar of foreachVars) {
            const newCombinations: string[][] = [];
            
            for (const existingCombination of combinations) {
                for (const value of foreachVar.values) {
                    const newCombination = [...existingCombination, value];
                    newCombinations.push(newCombination);
                }
            }
            
            combinations = newCombinations;
        }
        
        return combinations;
    }
}